### DISCUSSION TITLE
How do I parse localized date/time strings?

### DISCUSSION THREAD

**Discussion by willytrigg** [moment/luxon#1617] [Q&A]
Hello,

I am working on replacing Moment.js. One use case I have to replace is the parsing and validating of user-inputted dates in their browser's locale. For instance, the string '01/02/2024' should be interpreted as January 2, 2024 if the user has 'en' as their language or as February 1, 2024 in certain other locales. I believe there is another locale that has periods between the date units as well, so it should look for a string in the format '01.02.2024'.

The problem I have is that the 'fromFormat()' function takes in an exact format string/token rather than a localized one. For instance, it will interpret 'MM/dd/yyyy' as an exact format rather than a format that changes to suit the current locale. It appears that this library does have localized format strings/tokens, but they are only available for formatting, not parsing. 

Is there a way to get a localized parsing format dynamically? It is not possible for me to define a different format string for every locale that the browser supports, so I am not sure how to implement multi-language parsing.

Thanks

**Comment by diesieben07** [comment]
`fromFormat` supports several "macro tokens", that allow you to parse compound localized strings. How this works is that Luxon uses `Intl.DateTimeFormat#toLocaleString` to figure out how such a date looks in the given locale and then "reverse-engineers" this back into individual format tokens. It is not an exact science and parsing user-controlled input like this has [some drawbacks](https://moment.github.io/luxon/#/parsing?id=ad-hoc-parsing), but it usually works.
You can find the table of parsing tokens [here](https://moment.github.io/luxon/#/parsing?id=table-of-tokens). And example of a macro token would be `D` for a localized date like `01/02/2024`.
